File: src/main/java/com/memverge/splash/local/LocalTmpShuffleFile.java
package com.memverge.splash.local;

import com.memverge.splash.ShuffleFile;
import com.memverge.splash.TmpShuffleFile;
import lombok.val;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.UUID;

public class LocalTmpShuffleFile extends LocalShuffleFile implements TmpShuffleFile {
  private static final String TMP_FILE_PREFIX = "tmp-";
  private LocalShuffleFile commitTarget = null;

  private UUID uuid = null;
  private String targetId = null;
  private static final Logger log = LoggerFactory
      .getLogger(LocalTmpShuffleFile.class);

  private static final LocalTmpFileFolder tmpFolder = new LocalTmpFileFolder("local-tmp");
  public LocalTmpShuffleFile() {
    super();
  }

  public LocalTmpShuffleFile(String pathname)
      throws IOException {
    super(pathname);
  }

  public static LocalTmpShuffleFile make() throws IOException {
    //TODO:Done
    UUID uuid = UUID.randomUUID();
    String uuidStr = new File(folderPrefix(),
        String.format("%s%s", TMP_FILE_PREFIX, uuid.toString())).toString();
    LocalTmpShuffleFile ret = new LocalTmpShuffleFile(uuidStr);
    ret.create();
    ret.uuid = uuid;
    return ret;
  }

  public static LocalTmpShuffleFile make(ShuffleFile file) throws IOException {
    //TODO:Done
    if (file == null) {
      throw new IOException("Make null File as file");
    }
    LocalTmpShuffleFile ret = make();
    ret.commitTarget = (LocalShuffleFile) file;
    ret.targetId = file.getId();
    return ret;
  }

  @Override
  public void swap(TmpShuffleFile other) throws IOException {
    //TODO:Done
    if (!other.exists()) {
      String message = "Can only swap with a uncommitted tmp file";
      throw new IOException(message);
    }

    LocalTmpShuffleFile localTmpShuffleFileOther = (LocalTmpShuffleFile) other;
    if (localTmpShuffleFileOther == null) {
      throw new IllegalArgumentException("only accept LocalTmpShuffleFile instance");
    }

    forceDelete();
    UUID otherUuid = localTmpShuffleFileOther.uuid;
    String otherLocalId = localTmpShuffleFileOther.localId;
    localTmpShuffleFileOther.uuid = this.uuid;
    localTmpShuffleFileOther.localId = this.localId;
    this.uuid = otherUuid;
    this.localId = otherLocalId;
    File tmp = this.file;
    this.file = localTmpShuffleFileOther.file;
    localTmpShuffleFileOther.file = tmp;
  }

  @Override
  public LocalShuffleFile getCommitTarget() {
    //TODO:Done
    return this.commitTarget;
  }

  @Override
  public ShuffleFile commit() throws IOException {
    if (commitTarget == null) {
      throw new IOException("No commit target.");
    } else if (!exists()) {
      throw new IOException("Tmp file already committed or recalled.");
    }
    if (commitTarget.exists()) {
      log.warn("commit target already exists, remove '{}'.", commitTarget.getLocalId());
      commitTarget.forceDelete();
    }
    log.debug("commit tmp file {} to target file {}.",
        getLocalId(), getCommitTarget().getLocalId());

    rename(this.targetId);
    return commitTarget;
  }

  @Override
  public void recall() {
    //TODO:Done
    try {
      LocalShuffleFile commitTarget = getCommitTarget();
      if (commitTarget != null) {
        log.info("recall tmp file {} of target file {}.",
            getLocalId(), commitTarget.getLocalId());
      } else {
        log.info("recall tmp file {} without target file.", getLocalId());
      }
      delete();
    }catch (IOException e) {
        log.warn("Error while deleting '{}'.  Ignore it due to force.", localId);
    }
  }

  @Override
  public UUID uuid() {
    //TODO:Done
    return this.uuid;
  }

  public static String[] listAll() throws IOException {
    return new LocalShuffleFile(folderPrefix()).list();
  }

  public int getTmpFileCount() throws IOException {
    return this.listAll().length;
  }

  private static String folderPrefix() {
    return new File(tmpFolder.getTmpPath()).toString();
  }
}

File: src/main/scala/org/apache/spark/shuffle/local/LocalTmpShuffleFile.scala
/*
 * Copyright (C) 2018 MemVerge Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.spark.shuffle.local

import java.io._
import java.nio.file.Paths
import java.util.UUID

import com.memverge.splash.{ShuffleFile, TempFolder, TmpShuffleFile}
import org.apache.spark.internal.Logging

class LocalTmpShuffleFile extends TmpShuffleFile with Logging {

  private var _uuid: UUID = _
  private var commitTarget: LocalShuffleFile = _

  private def file: File = new File(LocalTmpShuffleFile.uuidToPath(uuid()))

  override def swap(other: TmpShuffleFile): Unit = {
    if (!other.exists) {
      val message = "Can only swap with a uncommitted tmp file"
      throw new IOException(message)
    }

    val otherLocal = other.asInstanceOf[LocalTmpShuffleFile]

    delete()

    val tmpUuid = otherLocal.uuid()
    otherLocal.setUuid(uuid())
    setUuid(tmpUuid)
  }

  private def setUuid(id: UUID): Unit = _uuid = id

  override def makeOutputStream(): OutputStream = {
    try {
      create()
      logDebug(s"create output stream for $getPath")
      new FileOutputStream(file, false)
    } catch {
      case e: FileNotFoundException =>
        throw new IllegalArgumentException(s"Create OS failed for $getPath.", e)
    }
  }

  override def getCommitTarget: ShuffleFile = commitTarget

  private[spark] def rename(tgtId: String): Unit = {
    val tgtFile = new File(tgtId)
    val tgtParent = tgtFile.getParentFile
    if (!tgtParent.exists()) {
      if (!tgtParent.mkdirs()) {
        logInfo(s"failed to create parent ${tgtParent.getAbsolutePath}")
      }
    }

    logDebug(s"rename $getPath to $tgtId.")
    LocalShuffleUtil.putBlock(file, tgtId)
  }

  override def commit(): ShuffleFile = {
    if (commitTarget == null) {
      throw new IOException("No commit target.")
    } else if (!exists) {
      create()
    }
    if (commitTarget.exists()) {
      logWarning(s"commit target already exists, remove '${commitTarget.getPath}'.")
      commitTarget.file.delete()
    }
    logDebug(s"commit tmp file $getPath to target file ${getCommitTarget.getPath}.")

    rename(commitTarget.getPath)
    commitTarget
  }

  override def recall(): Unit = {
    val commitTarget = getCommitTarget
    if (commitTarget != null) {
      logInfo(s"recall tmp file $getPath of target file ${commitTarget.getPath}.")
    } else {
      logInfo(s"recall tmp file $getPath without target file.")
    }
    delete()
  }

  override def uuid(): UUID = _uuid

  override def create(): TmpShuffleFile = {
    val parent = file.getParentFile
    if (!parent.exists()) {
      logInfo(s"create folder ${parent.getAbsolutePath}")
      parent.mkdirs()
    }
    logDebug(s"create file ${file.getAbsolutePath}")
    if (!file.createNewFile()) {
      throw new IOException(s"file $getPath already exists.")
    } else {
      logDebug(s"file $getPath created")
    }
    this
  }

  override def getSize: Long = file.length()

  override def delete(): Boolean = file.delete()

  override def exists(): Boolean = file.exists()

  override def getPath: String = file.getAbsolutePath

  override def makeInputStream(): InputStream = {
    try {
      log.debug("create input stream for {}.", getPath)
      new FileInputStream(file)
    } catch {
      case e: FileNotFoundException =>
        throw new IllegalArgumentException(s"Create IS failed for $getPath.", e)
    }
  }
}

private[spark] object LocalTmpShuffleFile {
  private val folder = TempFolder.getInstance()
  private val prefix = "tmp-"

  private def uuidToPath(uuid: UUID): String = {
    val tmpPath = folder.getTmpPath
    val filename = s"$prefix${uuid.toString}"
    Paths.get(tmpPath, filename).toString
  }

  def make(): LocalTmpShuffleFile = {
    val ret = new LocalTmpShuffleFile()
    ret.setUuid(UUID.randomUUID())
    ret
  }

  def make(file: ShuffleFile): LocalTmpShuffleFile = {
    require(file != null, "file should not be null")
    val ret = make()
    ret.commitTarget = file.asInstanceOf[LocalShuffleFile]
    ret
  }
}
